File: src/main/kotlin/com/clean/study/chapter03/java/AccountService.java
package com.clean.study.chapter03.java;

import com.clean.study.chapter03.Role;

public class AccountService{

    private  AccountStore accountStore;
    private static final String COMPANY = "clean code company";

    public AccountService(AccountStore accountStore) {
        this.accountStore = accountStore;
    }

    public Account create(String name , String phone, Role role) {
        Account alreadyPhoneAccount = accountStore.findByPhone(phone);
        if (alreadyPhoneAccount != null) {
            throw new IllegalArgumentException("duplicate phone");
        }
        Account account = null;
        switch(role) {
            case USER :
                account = new Account(name, phone, role);
                break;
            case OPERATOR :
                account = new Account(name, phone, role, COMPANY);
                break;
        }

        return accountStore.save(account);
    }

}

File: src/main/kotlin/com/clean/study/chapter03/AccountService.kt
package com.clean.study.chapter03

import java.lang.Exception
import java.lang.IllegalArgumentException

class AccountService(
    private val accountStore: AccountStore,
    private val accountFactoryImpl: AccountFactory
) {

    fun create(account: Account): Account {
        try {
            return createAccount(account)
        } catch (e: IllegalArgumentException) {
            logError(e)
            throw e
        }
    }

    private fun createAccount(account: Account): Account {
        checkAlreadyExistsPhone(account.phone)
        val account = accountFactoryImpl.makeAccount(account)
        return accountStore.save(account)
    }

    private fun checkAlreadyExistsPhone(phone: String) {
        accountStore.findByPhone(phone)?.run { throw IllegalArgumentException("duplicate phone") }
    }

    private fun logError(e: Exception) {
        println(e.message)
    }

}